core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -389,97 +528,129 @@ public long getUnpublishedWalTxnCount() {
     }
 
+    public TableToken getUpdatedTableToken(TableToken tableToken) {
+        return tableNameRegistry.getTokenByDirName(tableToken.getDirName());
+    }
+
     // For testing only
     @TestOnly
     public WalReader getWalReader(
-            CairoSecurityContext securityContext,
-            CharSequence tableName,
+            @SuppressWarnings("unused") CairoSecurityContext securityContext,
+            TableToken tableToken,
             CharSequence walName,
             int segmentId,
             long walRowCount
     ) {
-        if (tableSequencerAPI.hasSequencer(tableName)) {
-            // This is WAL table because sequencer exists
-            return new WalReader(configuration, tableName, walName, segmentId, walRowCount);
+        if (tableToken.isWal()) {
+            return new WalReader(configuration, tableToken, walName, segmentId, walRowCount);
         }
 
-        throw CairoException.nonCritical().put("WAL reader is not supported for table ").put(tableName);
+        throw CairoException.nonCritical().put("WAL reader is not supported for table ").put(tableToken);
     }
 
-    @Override
-    public @NotNull WalWriter getWalWriter(CairoSecurityContext securityContext, CharSequence tableName) {
+    @TestOnly
+    public @NotNull WalWriter getWalWriter(CairoSecurityContext securityContext, TableToken tableToken) {
         securityContext.checkWritePermission();
-        return walWriterPool.get(tableName);
+        verifyTableToken(tableToken);
+        return walWriterPool.get(tableToken);
     }
 
     public TableWriter getWriter(
             CairoSecurityContext securityContext,
-            CharSequence tableName,
+            TableToken tableToken,
             String lockReason
     ) {
         securityContext.checkWritePermission();
-        validNameOrThrow(tableName);
-        return writerPool.get(tableName, lockReason);
+        verifyTableToken(tableToken);
+        return writerPool.get(tableToken, lockReason);
     }
 
     public TableWriter getWriterOrPublishCommand(
             CairoSecurityContext securityContext,
-            CharSequence tableName,
+            TableToken tableToken,
             @NotNull AsyncWriterCommand asyncWriterCommand
     ) {
         securityContext.checkWritePermission();
-        validNameOrThrow(tableName);
-        return writerPool.getWriterOrPublishCommand(tableName, asyncWriterCommand.getCommandName(), asyncWriterCommand);
+        verifyTableToken(tableToken);
+        return writerPool.getWriterOrPublishCommand(tableToken, asyncWriterCommand.getCommandName(), asyncWriterCommand);
+    }
+
+    public TableWriter getWriterUnsafe(TableToken tableToken, String lockReason) {
+        return writerPool.get(tableToken, lockReason);
+    }
+
+    public boolean isTableDropped(TableToken tableToken) {
+        return tableNameRegistry.isTableDropped(tableToken);
+    }
+
+    public boolean isWalTable(TableToken tableToken) {
```

Comment:
we could just make TableToken::isWal() public and remove this method from CairoEngine.
it is more convenient, no need for a reference to the engine to call isWal().

Reply:
Yes, but I didn't want to make it public, doesn't feel right level of encapsulation to me. Token purpose is to identify the table, not provide table metadata.

Reply:
ok, fair enough